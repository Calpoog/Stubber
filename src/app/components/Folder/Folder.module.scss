@import '../../styles/colors.scss';

.folder {
  &.open {
    .wrapper:before {
      transform: rotate(-180deg);
    }
  }

  &.initiallyOpen {
    .stubs {
      overflow: visible;
      height: auto;
    }
  }

  &.disabled {
    .name {
      opacity: 0.5;
      text-decoration: line-through;
    }
  }
}

.header,
.wrapper {
  height: 60px;
}

.wrapper {
  font-weight: bold;
  cursor: pointer;
  position: relative;
  display: flex;
  align-items: center;

  & > :global(.actions) {
    opacity: 0;
    position: relative;
    right: -10px;
  }

  &:hover {
    & > :global(.actions) {
      opacity: 1;
    }

    &:before {
      opacity: 1;
    }
  }

  &:before {
    content: '';
    display: block;
    width: 13px;
    height: 8px;
    transform: rotate(0deg);
    transition: transform 0.2s ease-out;
    background-image: url(../../images/arrow.svg);
    opacity: 0.3;
    transition: 150ms;
  }

  &:global(.dragging) {
    position: fixed;
    top: 0;
    left: 10px;
    right: 10px;
    z-index: 1000;
    opacity: 0.5;
    width: auto;

    &,
    .handle {
      cursor: -webkit-grabbing;
    }
  }
}
.handle {
  position: absolute;
  width: 20px;
  height: 20px;
  top: 0;
  left: 0;
  z-index: 10;
  cursor: grab;
  user-select: none;
}
.content {
  padding: 0 15px;
  flex: 1;
}
.name {
  font-family: 'Roboto', sans-serif;
  color: $color-text;
  font-weight: bold;
  font-size: 16px;
  padding: 0;
  width: auto;
  background-color: transparent !important;
  border-color: transparent;
  position: relative;
  left: -1px;
}
.name[disabled] {
  cursor: pointer;
}
.stubs {
  overflow: hidden;
  height: 0;
  transition: height 0.2s ease-out;

  &:before {
    content: '';
    display: block;
    height: 1px;
    background: linear-gradient(90deg, rgba(42, 46, 62, 1) 0%, rgba(55, 61, 79, 1) 50%, rgba(42, 46, 62, 1) 100%);
  }
}
.stubsWrap {
  padding: 10px;
}
.emptyMessage {
  padding: 10px;
  opacity: 0.6;
}
