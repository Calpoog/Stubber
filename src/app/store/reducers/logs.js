export const REQUEST = 'REQUEST';
export const RESPONSE = 'RESPONSE';
export const CLEAR_LOG = 'CLEAR_LOG';

const initialState = { byID: [], byHash: {} };

export default function logs(state = initialState, action = {}) {
  switch (action.type) {
    case REQUEST: {
      return {
        byID: state.byID.concat(action.request.id),
        byHash: {
          ...state.byHash,
          [action.request.id]: action.request,
        },
      };
    }

    case RESPONSE: {
      const request = state.byHash[action.response.id];

      if (!request) return state;

      return {
        ...state,
        byHash: {
          ...state.byHash,
          [request.id]: {
            ...request,
            ...action.response,
          },
        },
      };
    }

    case CLEAR_LOG: {
      return initialState;
    }

    default:
      return state;
  }
}

export const addRequest = (request) => ({
  type: REQUEST,
  request,
});

export const addResponse = (response) => ({
  type: RESPONSE,
  response,
});

export const clearLog = () => ({
  type: CLEAR_LOG,
});

export const logActions = {
  addRequest,
  addResponse,
  clearLog,
};
