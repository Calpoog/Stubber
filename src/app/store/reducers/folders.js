import { ADD_STUB, REMOVE_STUB } from './stubs';
import { v4 as uuidv4 } from 'uuid';

export const ADD_FOLDER = 'ADD_FOLDER';
export const REMOVE_FOLDER = 'REMOVE_FOLDER';
export const EDIT_FOLDER = 'EDIT_FOLDER';
export const TOGGLE_FOLDER = 'TOGGLE_FOLDER';
export const MOVE_STUB = 'MOVE_STUB';
export const MOVE_FOLDER = 'MOVE_FOLDER';

export default function folders(state = { byID: [], byHash: {}, editing: null }, action = {}) {
  switch (action.type) {
    case ADD_FOLDER: {
      // also can add an array of f because it uses concat
      let byID = state.byID.slice(0);
      let byHash = { ...state.byHash };

      const folder = Object.assign({}, action.folder, {
        id: uuidv4(), // give new folder an ID
        stubs: [],
        open: true,
      });

      // add to the ordered ID list
      byID.unshift(folder.id);
      // add it to the hash map
      byHash[folder.id] = folder;

      return { byID, byHash, editing: folder.id };
    }

    case REMOVE_FOLDER: {
      const byHash = { ...state.byHash };
      delete byHash[action.folder.id];

      return {
        byID: state.byID.filter((id) => action.folder.id !== id),
        byHash,
      };
    }

    case EDIT_FOLDER: {
      const byHash = {
        ...state.byHash,
        [action.id]: {
          ...state.byHash[action.id],
          name: action.name,
        },
      };

      return { ...state, byHash, editing: null };
    }

    case TOGGLE_FOLDER: {
      const folder = state.byHash[action.id];
      const byHash = {
        ...state.byHash,
        [action.id]: {
          ...folder,
          open: !folder.open,
        },
      };

      return { ...state, byHash };
    }

    // handle ADD_STUB as well to put it in the right folder
    case ADD_STUB: {
      const folder = state.byHash[action.stub.folderID];
      const byHash = {
        ...state.byHash,
        [action.stub.folderID]: {
          ...folder,
          stubs: folder.stubs.concat(action.stub.id),
        },
      };

      return { ...state, byHash };
    }

    case MOVE_FOLDER: {
      const currentPosition = state.byID.indexOf(action.id);
      const position = action.position;

      if (currentPosition === position || currentPosition + 1 === position) {
        return state;
      } else {
        const byID = state.byID.slice(0);
        byID.splice(currentPosition, 1);
        byID.splice(position > currentPosition ? position - 1 : position, 0, action.id);
        return {
          ...state,
          byID,
        };
      }
    }

    case MOVE_STUB: {
      const byHash = { ...state.byHash };
      const oldFolder = byHash[action.oldFolderID];
      const newFolder = byHash[action.newFolderID];
      const stubID = action.stubID;
      let stubs = newFolder.stubs.slice(0);

      // if it changed folders, we remove it from the old one
      if (oldFolder.id !== newFolder.id) {
        byHash[oldFolder.id] = {
          ...oldFolder,
          stubs: oldFolder.stubs.filter((stubID) => stubID !== action.stubID),
        };
        // add it to the new one
        if (action.position === undefined) {
          stubs.push(stubID);
        } else {
          stubs.splice(action.position, 0, stubID);
        }
      } else {
        // otherwise we're moving it in its current folder
        const currentPosition = stubs.indexOf(stubID);
        // it moved if the position is not the original, or original + 1
        // because coming after itself is still its original position
        if (action.position !== currentPosition && action.position !== currentPosition + 1) {
          // if the new position is greater than current, subtract one to account for the moving of itself
          const newPos = action.position < currentPosition ? action.position : action.position - 1;
          stubs = stubs.filter((stubID) => stubID !== action.stubID);
          stubs.splice(newPos, 0, stubID);
        } else {
          // if it didn't move and the folder is the same, then nothing changed
          return state;
        }
      }

      // update the new/current folder
      byHash[newFolder.id] = {
        ...newFolder,
        stubs,
      };

      return { ...state, byHash };
    }

    case REMOVE_STUB: {
      const byHash = { ...state.byHash };
      const stub = action.stub;

      for (let i = 0; i < state.byID.length; i++) {
        const id = state.byID[i];
        const folder = byHash[id];

        if (folder.stubs.indexOf(stub.id) >= 0) {
          byHash[id] = {
            ...folder,
            stubs: folder.stubs.filter((id) => id !== stub.id),
          };
          break;
        }
      }

      return { ...state, byHash };
    }

    default:
      return state;
  }
}

export const addFolder = (folder) => ({
  type: ADD_FOLDER,
  folder,
});

export const removeFolder = (folder) => ({
  type: REMOVE_FOLDER,
  folder,
});

export const editFolder = (id, name) => ({
  type: EDIT_FOLDER,
  id,
  name,
});

export const toggleFolderOpen = (id) => ({
  type: TOGGLE_FOLDER,
  id,
});

export const moveStub = (oldFolderID, newFolderID, stubID, position) => ({
  type: MOVE_STUB,
  oldFolderID,
  newFolderID,
  stubID,
  position,
});

export const moveFolder = (id, position) => ({
  type: MOVE_FOLDER,
  id,
  position,
});

export const folderActions = {
  addFolder,
  removeFolder,
  editFolder,
  toggleFolderOpen,
  moveStub,
  moveFolder,
};

// gets ordered list of folders
const orderedFolders = (state) => state.folders.byID.map((id) => state.folders.byHash[id]);

// only recomputes when the order changes, but can't use reselect since the output function needs more state than input functions provide
export const getOrderedFolders = (function () {
  let cachedInput = undefined;
  let cachedOutput = undefined;
  return (state) => {
    let order = state.folders.byID;
    if (order !== cachedInput) {
      cachedInput = order;
      cachedOutput = orderedFolders(state);
    }

    return cachedOutput;
  };
})();
