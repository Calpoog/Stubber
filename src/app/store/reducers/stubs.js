import { IMPORT } from '../reducer';
import { MOVE_STUB, REMOVE_FOLDER } from './folders';
import { v4 as uuidv4 } from 'uuid';

export const ADD_STUB = 'ADD_STUB';
export const REMOVE_STUB = 'REMOVE_STUB';
export const EDIT_STUB = 'EDIT_STUB';
export const DISABLE_STUB = 'DISABLE_STUB';

// stubs state is just a hashmap of ID, ordering is taken care of on the folder level
export default function stubs(state = {}, action = {}) {
  switch (action.type) {
    case ADD_STUB: {
      const newState = { ...state };
      const stub = action.stub;

      // This mutates the action, because it is picked up by folders reducer and ID is needed
      stub.id = uuidv4(); // give new stubs an ID

      // add it to the hash map
      newState[stub.id] = stub;

      return newState;
    }

    case REMOVE_STUB: {
      const newState = { ...state };
      delete newState[action.stub.id];

      return newState;
    }

    case EDIT_STUB: {
      return {
        ...state,
        [action.stub.id]: {
          ...state[action.stub.id],
          ...action.stub,
        },
      };
    }

    case MOVE_STUB: {
      const { stubID, newFolderID } = action;
      return {
        ...state,
        [stubID]: { ...state[stubID], folderID: newFolderID },
      };
    }

    // handle remove folder because it should remove the associated stubs
    case REMOVE_FOLDER: {
      // this gets handled before folders reducer so the folder is still there, giving us the list of those
      // that need removed
      const stubs = { ...state };

      action.folder.stubs.forEach((id) => {
        delete stubs[id];
      });

      return stubs;
    }

    case DISABLE_STUB: {
      // can call using multiple IDs (for toggling entire folders)
      const stubs = { ...state };
      [].concat(action.id).forEach((id) => {
        stubs[id] = {
          ...stubs[id],
          disabled: action.disabled,
        };
      });

      return stubs;
    }

    default:
      return state;
  }
}

export const addStub = (stub) => ({
  type: ADD_STUB,
  stub,
});

export const removeStub = (stub) => ({
  type: REMOVE_STUB,
  stub,
});

export const editStub = (stub) => ({
  type: EDIT_STUB,
  stub,
});

export const disableStub = (id, disabled) => ({
  type: DISABLE_STUB,
  id,
  disabled,
});

export const stubActions = {
  addStub,
  removeStub,
  editStub,
  disableStub,
};
