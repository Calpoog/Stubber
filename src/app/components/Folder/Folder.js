import { useRef, useEffect, useLayoutEffect, memo } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { editFolder, moveFolder, removeFolder, toggleFolderOpen } from '../../store/reducers/folders';
import { disableStub } from '../../store/reducers/stubs';
import Stub from '../Stub';
import Actions from '../Actions';
import useDraggable from '../../hooks/draggable';
import { useHistory } from 'react-router';
import { useSelector, useDispatch } from 'react-redux';

import * as styles from './Folder.module.scss';

const Folder = memo(({ id }) => {
  const history = useHistory();
  const dispatch = useDispatch();
  const folder = useSelector((state) => state.folders.byHash[id]);
  const firstUpdate = useRef(true);
  const initiallyOpen = useRef(folder.open);
  const isEditing = useSelector((state) => state.folders.editing === id);
  const disabled = useSelector(({ folders, stubs }) => {
    const folder = folders.byHash[id];
    if (folder.stubs.length === 0) return false;

    for (let i = 0; i < folder.stubs.length; i++) {
      if (!stubs[folder.stubs[i]].disabled) return false;
    }
    return true;
  });

  const stubContainerElem = useRef(null);
  const stubWrapElem = useRef(null);
  const inputElem = useRef(null);

  const wrapperElem = useRef(null);
  const draggable = useDraggable('placer', wrapperElem, findPlacements, (placement) =>
    dispatch(moveFolder(folder.id, placement.index))
  );

  useEffect(() => {
    if (isEditing) editName();
  }, [isEditing]);

  useLayoutEffect(() => {
    if (!firstUpdate.current) {
      updateDropdown();
    } else {
      firstUpdate.current = false;
    }
  }, [folder.open]);

  function updateDropdown() {
    stubContainerElem.current.style = `height: ${
      !folder.open ? stubWrapElem.current.clientHeight : 0
    }px; overflow: hidden`;

    requestAnimationFrame(() => {
      stubContainerElem.current.style = `height: ${
        folder.open ? stubWrapElem.current.clientHeight : 0
      }px; overflow: hidden`;

      if (folder.open) setTimeout(() => (stubContainerElem.current.style = 'height: auto'), 200);
    });
  }

  function editName() {
    inputElem.current.disabled = false;
    inputElem.current.focus();
    inputElem.current.setSelectionRange(0, inputElem.current.value.length);
  }

  function saveName(e) {
    inputElem.current.disabled = true;
    dispatch(editFolder(folder.id, e.target.value));
  }

  function nameInput(e) {
    if (e.which === 13) inputElem.current.blur();
  }

  function addStub() {
    history.push(`/stub?folderID=${id}`);
  }

  function remove() {
    dispatch(removeFolder(folder));
  }

  function toggle() {
    dispatch(toggleFolderOpen(id));
  }

  function inputClick(e) {
    if (isEditing) e.stopPropagation();
  }

  function disable() {
    dispatch(disableStub(folder.stubs, !disabled));
  }

  console.log(`Folder ${id} render`);

  const folderClass = classNames('folder', 'surface', 'mb-3', styles.folder, {
    open: folder.open,
    [styles.open]: folder.open,
    [styles.initiallyOpen]: initiallyOpen.current,
    [styles.disabled]: disabled,
  });

  return (
    <div className={folderClass}>
      <div className={styles.header} onClick={toggle}>
        <div className={classNames('px-3', styles.wrapper)} ref={wrapperElem}>
          <div className={styles.handle} {...draggable}></div>
          <div className={styles.content} onClick={inputClick}>
            <input
              ref={inputElem}
              className={styles.name}
              type="text"
              onKeyPress={nameInput}
              onBlur={saveName}
              defaultValue={folder.name}
              disabled
            />
          </div>
          <Actions
            actions={{
              'Edit name': editName,
              'New stub': addStub,
              Remove: remove,
              [disabled ? 'Enable' : 'Disable']: disable,
            }}
          ></Actions>
        </div>
      </div>
      <div ref={stubContainerElem} className={styles.stubs}>
        <div className="stub-placement"></div>
        <div ref={stubWrapElem} className={styles.stubsWrap}>
          {folder.stubs.length === 0 && <div className={styles.emptyMessage}>No stubs.</div>}
          {folder.stubs.map((stubID) => (
            <Stub key={stubID} id={stubID} />
          ))}
        </div>
      </div>
    </div>
  );
});

Folder.propTypes = {
  id: PropTypes.string.isRequired,
};

function findPlacements() {
  const placements = [];
  document.querySelectorAll('.folder').forEach((folder, folderIndex) => {
    const box = folder.getBoundingClientRect();
    // one additional placement for before the first folder
    if (folderIndex === 0) placements.push({ y: box.top - 12, index: 0 });
    // placements are all below each folder
    placements.push({
      y: box.bottom + 8,
      index: folderIndex + 1,
    });
  });
  return placements;
}

export default Folder;
