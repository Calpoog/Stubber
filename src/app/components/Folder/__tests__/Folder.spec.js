import { shallow, mount } from 'enzyme';
import { useSelector, useDispatch } from 'react-redux';
import { act } from 'react-dom/test-utils';
import { TOGGLE_FOLDER } from '../../../store/reducers/folders';
import Folder from '../Folder.js';
import { DISABLE_STUB } from '../../../store/reducers/stubs.js';

const mockUseDispatch = jest.fn();

jest.mock('webextension-polyfill', () => ({
  runtime: { connect: jest.fn(() => ({ onMessage: { addListener: jest.fn() } })) },
}));
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn(() => mockUseDispatch),
}));

const folder = {
  name: 'Test Folder',
  id: 'calvin',
  open: true,
  stubs: [],
};
const defaultState = { folders: { byHash: { calvin: { ...folder, open: true } } } };
function mockStore(state = defaultState) {
  useSelector.mockImplementation((selector) => selector(state));
}

describe('Folder', () => {
  let wrapper;

  beforeEach(() => {
    mockStore();
    wrapper = shallow(<Folder id={'calvin'} />);
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  // it('renders', () => {
  //   expect(wrapper).toMatchSnapshot();
  // });

  it('shows name from props', () => {
    expect(wrapper.find('input').get(0).props.defaultValue).toEqual('Test Folder');
  });

  it('is open by default', () => {
    expect(wrapper.find('.folder').hasClass('open')).toBe(true);
  });

  it('is closed via state', () => {
    mockStore({ folders: { byHash: { asdf: { ...folder, open: false } } } });
    const component = shallow(<Folder id={'asdf'} />);
    expect(component.find('.folder').hasClass('open')).toBe(false);
  });

  it('closes the folder in the store', () => {
    wrapper.find('.header').simulate('click');
    expect(mockUseDispatch.mock.calls[0][0].type).toBe(TOGGLE_FOLDER);
  });

  // it('can start editing', () => {
  //   mockStore({ folders: { ...defaultState.folders, editing: 'calvin' } });
  //   let component;
  //   act(() => {
  //     component = mount(<Folder id={'calvin'} />);
  //   });
  //   console.log(component.find('input').getElement());
  //   expect(component.find('input').getDOMNode().disabled).toBe(true);
  // });

  // it('transitions to editing after click', () => {
  //   const component = shallow(<Folder key={1} id={0} folder={folder} editing={true} disabled={false} actions={{}} />);
  //   component.find('.edit').simulate('click');
  //   expect(component.find('.folder__name').getDOMNode().disabled).toBe(false);
  // });

  // it('saves edited name after input blur', () => {
  //   const component = shallow(
  //     <Folder key={1} id={0} folder={folder} editing={true} disabled={false} actions={actions} />
  //   );
  //   component.find('.edit').simulate('click');
  //   component.find('.folder__name').simulate('blur');
  //   expect(actions.editFolder).toBeCalled();
  // });

  it('is not disabled with no stubs', () => {
    expect(wrapper.find('.folder').hasClass('disabled')).toBe(false);
  });

  it('is not disabled when stubs are enabled', () => {
    mockStore({ folders: { byHash: { calvin: { ...folder, stubs: ['stub'] } } }, stubs: { stub: {} } });
    const component = shallow(<Folder id={'calvin'} />);
    expect(component.find('.folder').hasClass('disabled')).toBe(false);
  });

  it('is disabled via all stubs being disabled', () => {
    mockStore({ folders: { byHash: { calvin: { ...folder, stubs: ['stub'] } } }, stubs: { stub: { disabled: true } } });
    const component = shallow(<Folder id={'calvin'} />);
    expect(component.find('.folder').hasClass('disabled')).toBe(true);
  });
});
