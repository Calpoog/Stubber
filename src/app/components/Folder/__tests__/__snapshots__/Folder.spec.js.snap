// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Folder renders 1`] = `ShallowWrapper {}`;
